add mastodon package and module

petabyteboy opened the pull request:
###### Motivation for this change

This module can already be used to set up a fully functional mastodon instance, but there is still a lot to do:

- [x] Discuss if moretea/yarn2nix should be added to nixpkgs again
- [x] Don't force users to store secrets in nix store
- [x] Improve documentation of module options
- [x] Add mastodon user and group ids
- [x] Add meta information to the mastodon package
- [ ] Investigate if it makes sense to allow enabling and disabling the three services seperately, similar to the kubernetes module

Some things would be nice to have but are not strictly required for a first version in my opinion:

- [x] Write some tests
- [ ] Investigate if building streaming, sidekiq and web services seperately would be possible and advantageous
- [ ] Add more advanced options, i.e. the number of sidekiq threads, support for S3 storage backend, ...
- [x] Package mastodon tools like tootctl in a way that makes them easy to use

Any feedback is appreciated :-)

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nix-review --run "nix-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [x] Assured whether relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

Commit:
mastodon: init at 3.0.1

Commit:
nixos/mastodon: init

nixos/mastodon: add webserver test
nixos/mastodon: add package tests
Tests and fixes contributed by @happy-river, thanks!

petabyteboy commented:
Of course we are facing the same IFD issue with yarn2nix as with riot-desktop and #59111.
It seems like the only solution for now is including yarn.nix and package.json in nixpkgs.

petabyteboy commented:
Can I do ...? @GrahamcOfBorg eval

petabyteboy commented:
Can I do ...?
@GrahamcOfBorg eval

petabyteboy commented:
Unfortunately it seems like this is a dead end. I will try to use https://github.com/Profpatsch/yarn2nix tomorrow.

alyssais commented:
> Can I do ...?
> @GrahamcOfBorg eval

As I understand it, you should be able to eval, but not build, unless
you're a known or trusted user, in OfBorg terminology.

petabyteboy commented:
> Can I do ...? @GrahamcOfBorg eval
> As I understand it, you should be able to eval, but not build, unless you're a known or trusted user, in OfBorg terminology.

That's what I concluded too, but as it says in the OfBorg readme, there is no reason to call eval manually since it happens automatically.

petabyteboy commented:
After trying to integrate profpatsch/yarn2nix into nixpkgs, I give up on that. There are multiple broken Haskell packages required for profpatsch/yarn2nix to run.

My new plan is to add support for Hydra-enabled git dependencies to moretea/yarn2nix.

petabyteboy commented:
Hooray /o/

aanderse commented on nixos/modules/services/web-apps/mastodon.nix:
```diff
@@ -0,0 +1,333 @@
+      };
+
+      dbUser = lib.mkOption {
+        description = "Postgres login username";
```
Is it much work to add a `dbType` option and support mysql as well?

petabyteboy replied:
I think this is not supported by mastodon.

aanderse replied:
Ah, ok.

aanderse reviewed (commented):
I left a few comments about the nixos module which I hope you find helpful. Feel free to use or ignore as you see fit.

petabyteboy commented on nixos/modules/misc/ids.nix:
```diff
@@ -339,6 +339,7 @@
       rss2email = 312;
       cockroachdb = 313;
       zoneminder = 314;
+      mastodon = 315;
```
I'm not sure if it's necessary. In what cases is it usually necessary?

aanderse replied:
Generally it is not required, though many modules in nixos have unnecessarily set one.

petabyteboy commented on nixos/modules/services/web-apps/mastodon.nix:
```diff
@@ -0,0 +1,333 @@
+  config = lib.mkIf cfg.enable {
+    systemd.services.mastodon-init-dirs = {
+      script = ''
+        mkdir -p ${tmpDir}
```
Even better I will use PrivateTmp in the service

petabyteboy commented:
Thanks, I think I have implemented all your suggestions.

zimbatm commented:
I don't know of a good way to add yarn2nix without forking the code base. When we had it in the tree it started diverging quickly with both sides receiving patches. Before adding it back we need to decide on how to resolve that issue. A few possibilities:

1. Add yarn2nix as a git submodule to nixpkgs, probably doesn't work with hydra
2. Use `git substree` to keep yarn2nix in sync. This needs a bit of a wrapper scripts action.
3. Archive upstream, declare nixpkgs to hold the canonical version.
4. ?

petabyteboy commented:
I have another suggestion. 

> I don't know of a good way to add yarn2nix without forking the code base. When we had it in the tree it started diverging quickly with both sides receiving patches. Before adding it back we need to decide on how to resolve that issue. A few possibilities:
> 
> 1. Add yarn2nix as a git submodule to nixpkgs, probably doesn't work with hydra
> 2. Use `git substree` to keep yarn2nix in sync. This needs a bit of a wrapper scripts action.
> 3. Archive upstream, declare nixpkgs to hold the canonical version.
> 4. ?

I agree it has to be discussed and well-thought-through. 
Of the solutions you suggested, I prefer 3., but I have another idea:

Actually yarn2nix can be split into two parts, similar to bundix and bundlerEnv. 
- One part is responsible for generating the nix expression and is maintained in a seperate repository. This command line tool can be provided in nixpkgs just like all other tools with fetchFromGitHub. 
- The other part contains everything for building a node_modules folder from the generated nix expression and is kept in nixpkgs.

nixos-discourse commented:
This pull request has been mentioned on **Nix community**. There might be relevant details there:

https://discourse.nixos.org/t/has-anyone-ran-mastodon-on-nixos/1440/9

nyanloutre commented:
Thanks for the work, I will test this on my server as soon as possible

alyssais commented:
> > I don't know of a good way to add yarn2nix without forking the code base. When we had it in the tree it started diverging quickly with both sides receiving patches. Before adding it back we need to decide on how to resolve that issue. A few possibilities:
> >
> > 1. Add yarn2nix as a git submodule to nixpkgs, probably doesn't work with hydra
> > 2. Use `git substree` to keep yarn2nix in sync. This needs a bit of a wrapper scripts action.
> > 3. Archive upstream, declare nixpkgs to hold the canonical version.
> > 4. ?
>
> I agree it has to be discussed and well-thought-through.
> Of the solutions you suggested, I prefer 3., but I have another idea:
>
> Actually yarn2nix can be split into two parts, similar to bundix and bundlerEnv.

I like the sound of this.

nyanloutre commented:
My server is on 19.03 so I will need a bit of work to backport this it seems

petabyteboy commented:
> My server is on 19.03 so I will need a bit of work to backport this it seems

It shouldn't be too much, if you have any questions let me know

lilyball commented on nixos/modules/services/web-apps/mastodon.nix:
```diff
@@ -0,0 +1,330 @@
+    services.mastodon = {
+      enable = lib.mkEnableOption "Mastodon federated social network";
+      configureNginx = lib.mkOption {
+        description = "Use nginx as reverse proxy for mastodon";
```
It would be great to have some documentation here of what the user has to do if they want to disable this and use a different reverse-proxy (such as Caddy).

petabyteboy replied:
It's actually disabled by default.

petabyteboy replied:
It's actually not so easy to tell if it should be enabled or disabled by default. If it stays enabled by default, I think we have to add a check that mastodon itself is enabled before setting the options.
I thought I had left it disabled by default, but now that I think about it it should probably stay enabled, because the service is not really usable without a correctly configured reverse proxy.

lilyball replied:
I would be pretty surprised if I went to enable mastodon and ended up with nginx running without warning, given that I already have Caddy set up to listen on ports 80/443 and nginx would try to use those ports as well, which wouldn't work. I'd prefer to have this default to `false` with appropriate documentation.

It's also worth pointing out that the nginx configuration used here relies on the structured vhosts, and if I have nginx configured using a custom config file instead then it wouldn't work.

Mastodon may only document how to use nginx, but the actual proxy requirements can be inferred from the nginx config file and replicated using other reverse proxies. Even if you don't document the exact requirements here, you could link to the Mastodon documentation on how to set up the reverse proxy.

petabyteboy replied:
As you wish, let's see if anyone else complains about it being disabled by default later. 
I'm just wondering where I should add that documentation.

petabyteboy commented:
> My server is on 19.03 so I will need a bit of work to backport this it seems

I just tried to build the package with 19.03. You only have to cherry-pick https://github.com/NixOS/nixpkgs/pull/60693/ and somehow the hashes of the two ruby dependencies that are pulled from git changed depending on what version of nixpkgs I used, so I needed to change those.

kampka commented:
Can you please make sure the state directory is configurable?
It would help immensely for keeping control over user generated content, such as avatars, media, etc. which needs to be backed up regularly.
A typical name for this option in nix services is `dataDir` which could easily default to `/var/lib/mastodon`.

Thanks for your effort. :+1:

aanderse commented:
> Can you please make sure the state directory is configurable?
> It would help immensely for keeping control over user generated content, such as avatars, media, etc. which needs to be backed up regularly.
> A typical name for this option in nix services is `dataDir` which could easily default to `/var/lib/mastodon`.
> 
> Thanks for your effort. +1

@kampka can you please elaborate on what backup and control needs you have that won't work with `/var/lib/mastadon`?

lilyball commented on nixos/modules/services/web-apps/mastodon.nix:
```diff
@@ -0,0 +1,339 @@
+  options = {
+    services.mastodon = {
+      enable = lib.mkEnableOption "Mastodon federated social network";
+      configureNginx = lib.mkEnableOption ''
```
This should be `lib.mkOption`, because the argument to `mkEnableOption` is the name, meaning the description of this option is going to be

```
Whether to enable Configure nginx as reverse proxy for mastodon
Alternatively you can configure a reverse-proxy of your choice to serve these paths:
…
https://github.com/tootsuite/mastodon/blob/master/dist/nginx.conf.
```

petabyteboy replied:
I'm a bit confused, where should this text go instead? There is only one description/name for one option or is there a separate attribute for an extended description?
Until now it was intended to describe the alternative in the description of configureNginx, since I didn't know where to put it.

lilyball replied:
The problem with `mkEnableOption` is it tries to synthesize a description from a name, but when you're providing a full description, that doesn't particularly work. I'm suggesting

```nix
configureNginx = lib.mkOption {
  default = false;
  example = true;
  description = ''
    Configure nginx as reverse proxy for mastodon
    Alternatively you can configure a reverse-proxy of your choice to serve these paths:
    …
    https://github.com/tootsuite/mastodon/blob/master/dist/nginx.conf
  '';
  type = lib.types.bool;
};
```

kampka commented:
> @kampka can you please elaborate on what backup and control needs you have that won't work with `/var/lib/mastadon`?
 
Sure.
1. On systems I maintain, `/var/lib` is typically not located on a large partition. Separate partitioning enables flexible backup strategies like snapshotting dedicated partitions, enables better security mount options (noexec, nodev, etc.), makes storage easier, eg. prevents root from running out of space because someone uploaded to may gifs, etc.

2. `/var/lib/mastodon` is a magic string that can't be referenced in other services and configurations, eg. backup and monitoring. This makes it's use quite inconvenient and is in my opinion not idiomatic for a nix module. Basically, it complicates a lot of things down the line for no gain whatsoever.

aanderse commented:
@kampka Is there any reason using a symlink for your first point wouldn't work?

The idiomatic behaviour you mention in nixos modules comes with many problems itself. See https://github.com/NixOS/nixpkgs/pull/56265#discussion_r267738399 for a brief discussion on why adding a `dataDir` is inconvenient.

cc @flokli and @Infinisil as possibly interested parties

petabyteboy commented:
I must say I would appreciate the option to set the data storage path as well. Similar to @kampa I have a seperate data disk that is not mounted at /var/lib where I want the data to be stored. My solutions for now were symlinks or bind-mounts, and it mostly worked fine, but where a dataDir option is provided I prefer to use that.

kampka commented:
> The idiomatic behaviour you mention in nixos modules comes with many problems itself. See [#56265 (comment)](https://github.com/NixOS/nixpkgs/pull/56265#discussion_r267738399) for a brief discussion on why adding a `dataDir` is inconvenient.

Left a short comment on that issue about that.

> Is there any reason using a symlink for your first point wouldn't work?

A symlink would work fine, but I don't think it addresses any of the issues.

* You would still have to have a service unit with specific permissions to create that symlink. This service already has a setup unit `mastodon-init-dirs` that creates all necessary state which could create such a symlink as well but that would be kind of strange if you pass a path into the configuration just to have it symlinked to when you could just use it directly.

* You would also still have to _know_ the magic strings to know where data is stored for different services, which can change under the hood. Iff there was something like a known state dir table similar to the known user id tabel in the Nix path, this point could be mitigated, but it still leaves the first point.

nyanloutre commented:
I still don't understand why using a standard directory is such a problem. `/var/lib` is not a magic string, it's the standard directory used by the `StateDirectory` option. You should expect to see all application states in this directory.

kampka commented:
> I still don't understand why using a standard directory is such a problem. `/var/lib` is not a magic string, it's the standard directory used by the `StateDirectory` option. You should expect to see all application states in this directory.

I really don't want to debate the meaning of the word _standard_ here, especially on the Nix issue tracker, as it is beside the point.
The point is, I do not want to have to guess where non-volatile data is located on the systems I maintain, reasons are stated above.
One of the main benefit of NixOS, at least for me, is that I do not have to drag magic constants around in my configuration (port numbers, directories, user ids, etc.). Instead, I can reference them in my derivations an be done with the whole problem, changes just converge.
If I now have to have a separate lookup table of paths that may or may not be used anymore without me noticing when the nix expressions changes, this feature becomes quite useless because now I have to _know_ these things again.

alyssais commented:
I’m pro `dataDir`. This thread shows us that users clearly value being able to choose where to store their state, and I don’t think it’s fair to force them to mess around with symlinks or bindmounts because of a systemd misfeature.

alyssais commented:
That said, if it’s difficult to implement I wouldn’t worry about doing it _now_. But in general I think it would be a good idea.

aanderse commented:
I'll state again what I originally said to @petabyteboy:

> I left a few comments about the nixos module which I hope you find helpful. Feel free to use or ignore as you see fit.

The `dataDir` option is definitely a pattern in nixos modules and anyone is free to use it (or not use it) as they see fit. I didn't mean to turn this PR into the battleground of those for and against the option.

Thanks for everyone's opinion up to this point, maybe I'll make a separate issue where the topic can be discussed further.

aanderse commented:
So 29 days later... where are we with this? I'm under the impression having `mastadon` in nixpkgs is of high value to more than just a few users.

nixos-discourse commented:
This pull request has been mentioned on **Nix community**. There might be relevant details there:

https://discourse.nixos.org/t/prs-ready-for-review-may-2019/3032/13

ashkitten commented:
i'd definitely like to see an option to run custom versions of mastodon; many instances are customized in at least a few ways, some run forks like [glitch-soc](https://github.com/glitch-soc/mastodon)

leo60228 commented:
To clarify, if I clone this fork on my server, I could host a Mastodon instance today?

petabyteboy commented:
> To clarify, if I clone this fork on my server, I could host a Mastodon instance today?

Yes, I am running my mastodon instance social.pbb.lc using this module. I am using my overlay at https://git.pbb.lc/petabyteboy/nixfiles/src/branch/master/pkgs/mastodon instead of a nixpkgs fork though.

petabyteboy commented:
Short summary on the status:

- It works. You can run your own instance using my overlay from https://git.pbb.lc/petabyteboy/nixfiles/
- Included tools like tootctl do not work out of the box, but are not strictly needed to run an instance.
- Some more general discussion is needed on the following topics before mastodon can be included in nixpkgs:
  - How do we handle yarn2nix in nixpkgs
  - Do we want to merge this without some solution for yarn.nix and gemset.nix (IFD or recursive nix)

matthiasbeyer commented:
Any progress?

matthiasbeyer commented on nixos/modules/services/web-apps/mastodon.nix:
```diff
@@ -0,0 +1,339 @@
+          Path to file containing the OTP secret
+
+          Can be generated by running
+          cd $(nix-instantiate --eval '&lt;nixpkgs&gt;' -A mastodon.outPath); bin/rake secret
```
Is this really the same as for `secretKeyBaseFile`?

petabyteboy replied:
Yes. The secrets rake task generates an OTP secret and a base secret.

ashkitten commented:
is there no documentation on how to update the package?

ashkitten commented:
i played around with this myself today, and i believe it needs quite a few modifications before it's actually usable:

- the nginx config that comes with it does not use `recommendedProxySettings`, which is required so that mastodon won't just continually redirect to https://127.0.0.1
- it's not clear that you will need to set up postgresql separately
- by default it uses postgresql via a local ip instead of the unix socket, which requires that a password be set
- it does not set up the database for you and there is no indication that you will have to set up the database yourself
- a lot of options are required to use the service at all, which are not actually necessary - we can make .env settings optional.
- putting each secret in a separate file is very confusing, especially since they are all lumped into a single file afterwards anyway
- when changing one of the secret variables, the secrets file was not regenerated. i couldn't figure out how to get it to regenerate, and ended up having to edit the secrets file manually
- i set it up with my personal fork of mastodon, which ended up being a real pain to figure out how to update dependencies. having to include both the `package.json` and `yarn.lock` as well as generate `yarn.nix` and `gemset.nix` was a nightmare, especially since i didn't know i had to apply `mastodon-nix.patch` before using `bundix` and `yarn2nix` to generate those files. needs better workflow and documentation.

petabyteboy commented:
Thanks for the review. Do you have any idea how we can improve this?

aanderse commented:
As far as the database goes maybe you can lump all database options under a `database` attribute like many other services do and then eventually take advantage of [this](https://github.com/NixOS/nixpkgs/issues/62073) when it gets done.

FYI there are `services.postgresql.enensureDatabases` and ` services.postgresql.ensureUsers` options in master that will provision a database and database user for you, which combine nicely with a `database.createLocally` option. Take a look at [zabbix](https://github.com/NixOS/nixpkgs/blob/master/nixos/modules/services/monitoring/zabbix-server.nix) in master for an example of working with databases. Bonus points for supporting multiple database types if the application allows it :laughing:

ashkitten commented:
for the secrets, i don't see any point in having some of the env variables separate from the rest. i think we should just have an `envFile` option, which also solves the issue of the user wanting to specify things that aren't configurable in this module. however, we still need the domain name for configuring nginx if we still want to do that. i suppose we can keep the domain as a configuration option in the module, but we have to decide whether that will set the domain for mastodon or just nginx.

petabyteboy commented:
> for the secrets, i don't see any point in having some of the env variables separate from the rest. i think we should just have an `envFile` option, which also solves the issue of the user wanting to specify things that aren't configurable in this module. however, we still need the domain name for configuring nginx if > we still want to do that. i suppose we can keep the domain as a configuration option in the module, but we have to decide whether that will set the domain for mastodon or just nginx.

I disagree about the env file: The env file contains all configuration options. Managing those configuration options is the purpose of the module after all. Managing those values in nix makes setting and overriding options much easier.
For the env file not always being regenerated, this can be solved by documentation or maybe we can adjust the systemd units to regenerate the configuration on every start of the mastodon service.

ashkitten commented:
in that case we definitely need an `extraEnv` and `extraEnvFile` option in the module to make it possible to specify other variables and possibly other secrets.

petabyteboy commented:
I will update to 2.9.2 and implement the suggestions (changes for database options, extraEnv) soon :-)

petabyteboy commented:
> i played around with this myself today, and i believe it needs quite a few modifications before it's actually usable:
> 
> * the nginx config that comes with it does not use `recommendedProxySettings`, which is required so that mastodon won't just continually redirect to https://127.0.0.1
> * additionally, nginx needs to have SSL, as recent versions of mastodon do not support http in production.
> * it's not clear that you will need to set up postgresql separately
> * by default it uses postgresql via a local ip instead of the unix socket, which requires that a password be set
> * it does not set up the database for you and there is no indication that you will have to set up the database yourself
> * a lot of options are required to use the service at all, which are not actually necessary - we can make .env settings optional.
> * putting each secret in a separate file is very confusing, especially since they are all lumped into a single file afterwards anyway
> * when changing one of the secret variables, the secrets file was not regenerated. i couldn't figure out how to get it to regenerate, and ended up having to edit the secrets file manually
> * i set it up with my personal fork of mastodon, which ended up being a real pain to figure out how to update dependencies. having to include both the `package.json` and `yarn.lock` as well as generate `yarn.nix` and `gemset.nix` was a nightmare, especially since i didn't know i had to apply `mastodon-nix.patch` before using `bundix` and `yarn2nix` to generate those files. needs better workflow and documentation.

On updating to a new version or using a forked version, I think we might be able to upstream some of our changes if upstream allows it. Also note that some of the features in the nix tooling are not merged yet (https://github.com/moretea/yarn2nix/pull/100, https://github.com/manveru/bundix/pull/48), which makes it difficult to generate yarn.nix and gemset.nix.

ashkitten commented:
it would be really great if we could generate yarn.nix and gemset.nix on the fly, and then we would be able to simply specify what repository we want to use as a module option.

petabyteboy commented:
> it would be really great if we could generate yarn.nix and gemset.nix on the fly, and then we would be able to simply specify what repository we want to use as a module option.

Unfortunately this is not possible due to Hydra restrictions.

ashkitten commented:
> > it would be really great if we could generate yarn.nix and gemset.nix on the fly, and then we would be able to simply specify what repository we want to use as a module option.
> 
> Unfortunately this is not possible due to Hydra restrictions.

i'm unclear, what is hydra unable to do?

petabyteboy commented:
yarn2nix natively supports generating yarn.nix on the fly, but it requires IFD (import from derivation) functionality, which is disabled in Hydra builds. When you import .nix files from previous build outputs, that's IFD. All .nix files interpreted by Hydra need to be part of nixpkgs.

petabyteboy commented:
Another restriction with gemset.nix is that the tool generating it requires an internet connection, because the Gemfile/Gemfile.lock doesn't contain usable hashes. So while yarn.nix contains the same information as yarn.lock (and can be generated without an internet connection), gemset.nix contains additional information pulled from the internet (the hashes of all dependencies).
Even for yarn.nix there is another problem: Git dependencies don't have a hash in yarn.lock, the builtin fetchGit function doesn't work on Hydra, and fetchgit requires a hash again which is not contained in the mastodon sources.

I have spent a lot of time thinking through this and we will always need a manual step outside the build to prepare the mastodon source.

ashkitten commented:
hmm, that's unfortunate. if it can't be generated during build, perhaps we should have a maintainer script at least which prepares everything to be built

petabyteboy commented:
I have another idea on how we can make it easier to use a custom mastodon source. When overriding the mastodon source, the package is built locally, so the Hydra restrictions don't apply. 
There is still the problem with gemset.nix and yarn.nix containing extra information that needs to be fetched from the internet, but this could be worked around with a fixed-output derivation, although I think someone was discouraging this.

ashkitten commented:
not sure how you mean to have gemset.nix and yarn.nix generated during build, but only if the source is overridden

ashkitten commented:
i would feel bad about merging this if there isn't first-class support for custom forks of mastodon. to be honest, if we can't provide a module option to let the user easily specify a custom source i'd rather provide this as an overlay in a separate repository

spacekookie commented:
@ashkitten:
> * it's not clear that you will need to set up postgresql separately

Isn't that default behaviour for most modules? There's other modules (such as `quassel`, I believe) where you need to make sure postgres is explicitly configured to make it work.

Apart from that, I don't feel this is a bad thing, especially considering that database setup is something that people are gonna have feelings about. In fact, I'd go as far as saying that it'd make me feel a bit weird if the `mastodon` module did this for me.

aanderse commented:
I think one goal of NixOS module authors should always be to make as little configuration work for the user as possible so that a normal user of a software who just doesn't care how things are implemented can type `services.foo.enable = true;` and not too much more to have a working service.

I think another goal should be to allow advanced users a reasonable amount of customization so they aren't locked in with generic options that the module author wrote.

Many modules automatically provision databases and most of those have a `database.createLocally` option which defaults to`true`. Advanced users should feel free to set that to false because they know what they want and how to achieve it.

petabyteboy commented:
I have never used such a module that creates a database for me by default. Thanks for explaining how some modules do it @aanderse! I see how this would be a good addition, but is it actually required for all new modules? I know a number of modules that don't set up all required services. If it is not a requirement, I would like to add support for `database.createLocally` later.

aanderse commented:
@petabyteboy Definitely not a requirement! Just a nice to have :smile:

@spacekookie Good to know about `quassel`. If I find the time I'll try and add a `database` option to it.

petabyteboy commented:
I added an update script, which can be used to generate all necessary files for a new release or when using a fork. 
I updated the package to the latest master of mastodon for now, because it contains https://github.com/tootsuite/mastodon/pull/11444 and I don't want to put effort into working around something that will be gone in a few weeks.
At some point I will try to redo the packaging part to be much more clean, by just copying what's needed for every service, and making all the executables work properly (tootctl, ...).

Infinisil commented on nixos/modules/services/web-apps/mastodon.nix:
```diff
@@ -0,0 +1,339 @@
+    systemd.services.mastodon-init-db = {
+      script = ''
+        cd ${pkgs.mastodon}
+        rake db:migrate
```
Note that if this does automatic database migration, NixOS rollbacks might not work

Infinisil replied:
I've been wondering myself how this should be handled best but haven't come to a good conclusion

petabyteboy replied:
I think this is a question of state management, which is not really dealt with by Nix. The same problem exists for all stateful data on the system (it can not be rolled back unless the user has a backup). Maybe a note in the docs would be good to remind users of that.

petabyteboy replied:
As soon as tootctl works, I will add a "cache clear" call. This should be sufficient for 99% of the releases. The mastodon releases page lists the following steps for a major release:
 - Install dependencies - done by Nix
 - Run database migrations - done by mastodon-init-db.service
 - Precompile the assets - done by Nix
 - Restart all Mastodon processes - done by the activate script
 - Clear cache - this is what we are still missing

petabyteboy replied:
I thought the migrate task worked fine for an empty database on my deployment. I would say I will take a look but I'm pretty busy. If it doesn't work for an empty database, we will of course need that.

alyssais replied:
Migrating an empty DB will be slow and error prone, if nothing else, since it has to construct it migration by migration and some of those migrations may not have been touched for years.

Miaourt replied:
I'm really uneasy about doing migrations automatically, even more when no automatic backups is done before.

I wish there could be an option to disable it.
(Btw, thanks for your work <3 !)

petabyteboy replied:
Sure, I can add an option to opt-out.

petabyteboy replied:
added an automaticMigrations option

happy-river commented on pkgs/servers/mastodon/version.nix:
```diff
@@ -0,0 +1 @@
+"v2.9.3"
```
```suggestion
"2.9.3"
```
For consistency with the rest of Nixpkgs.

petabyteboy replied:
There are some packages with and some without v in nixpkgs, depending on the upstream tags. In the case of mastodon the upstream tags include a v. If you still think it should be changed, please also change the update script to remove the v when updating the release.

alyssais replied:
Agree this shouldn’t have the v.

petabyteboy replied:
done

happy-river commented:
I've written [a test for this module](https://github.com/happy-river/nixpkgs/tree/mastodon-test) which verifies that the web server comes up.  Feel free to incorporate it into this pull request. I plan on writing another test which runs Mastodon's test suite, which needs to be run in a NixOS test because it requires postgresql and redis.

petabyteboy commented on pkgs/servers/mastodon/default.nix:
```diff
@@ -0,0 +1,77 @@
+  };
+
+  mastodon-js-modules = yarn2nix-moretea.mkYarnPackage {
+    name = "mastodon-modules";
```
Does yarn2nix like that?

alyssais commented on nixos/modules/services/web-apps/mastodon.nix:
```diff
@@ -0,0 +1,339 @@
+
+  env = {
+    RAILS_ENV = "production";
+    NODE_ENV = "production";
```
Mild disagree. I’m not sure running Mastodon through a NixOS-generated systemd service that has to be rebuilt every time you make a change is a reasonable development flow.

flokli replied:
Shouldn't a user be able to manually do that via
```nix
systemd.services.mastodon-init-db.environment.RAILS_ENV = lib.MkForce "development";
systemd.services.mastodon-streaming.environment.RAILS_ENV = lib.MkForce "development";
systemd.services.mastodon-web.environment.RAILS_ENV = lib.MkForce "development";
systemd.services.mastodon-sidekiq.environment.RAILS_ENV = lib.MkForce "development";
```
(or a subset of that) anyways?

petabyteboy replied:
iI'm not sure if this would have a higher priority than EnvironmentFile.
But now that you mention it, I would like to expose the environment variables as a NixOS option.

petabyteboy replied:
You can now use `services.mastodon.extraOptions.RAILS_ENV = "development";`

alyssais commented on pkgs/servers/mastodon/default.nix:
```diff
@@ -0,0 +1,77 @@
+      rm -f bin/$b
+      ln -s ${mastodon-gems}/bin/$b bin/$b
+    done
+
```
> It seems unlikely that overwriting them with executables from the bundled environment is going to work correctly.

Can we confirm this before complicating the code? Fwiw, we can always use patchShebangs rather than doing send.

petabyteboy replied:
Just doing patchShebangs does not work for all commands.

happy-river replied:
As a specific example, if you overwrite `bin/rails` with the `rails` from the bundled environment, then the command `rails db:migrate` which is suggested in [the Mastodon documentation](https://docs.joinmastodon.org/administration/troubleshooting/) does not work.  It does work if you use Mastodon's `bin/rails` script with the path to Ruby patched.  My previous suggestion did not patch all the instances of `#!/usr/bin/env ruby` in the `bin` directory, so here is an improved version:

```
    sed -i 's#/usr/bin/env ruby#${mastodon-gems.wrappedRuby}/bin/ruby#g' bin/*

    for b in $(ls ${mastodon-gems}/bin/)
    do
      if [ ! -f bin/$b ]; then
        ln -s ${mastodon-gems}/bin/$b bin/$b
      fi
    done
```

alyssais replied:
What happens with `patchShebangs`?

petabyteboy replied:
> What happens with `patchShebangs`?

The ruby shebang gets replaced and the script can be executed, but it fails to find any dependencies from mastodon-gems.

petabyteboy replied:
> The executables in Mastodon's `bin` directory are scripts with `/usr/bin/env ruby` shebangs which load configuration from elsewhere in the source tree before doing whatever it is they do. It seems unlikely that overwriting them with executables from the bundled environment is going to work correctly. I suggest:
> 
> ```
>     for b in $(ls ${mastodon-gems}/bin/)
>     do
>       if [ -f bin/$b ]; then
>         sed -i 's#/usr/bin/env ruby#${mastodon-gems.wrappedRuby}/bin/ruby#g' bin/$b
>       else
>         ln -s ${mastodon-gems}/bin/$b bin/$b
>       fi
>     done
> ```

Even with this, tootctl fails with
```
Could not find rake-12.3.3 in any of the sources
Run `bundle install` to install missing gems.
```
To make tootctl work I had to combine the first few lines of one of the ruby scripts in mastodon-gems/bin/ with the tootctl script:

```
#!/nix/store/jcamgkl7k805gzvw8v5h9fsxi2nidqnv-ruby-2.6.5/bin/ruby

ENV["BUNDLE_GEMFILE"] = "/nix/store/fwpr79frmm1iiz4nr057fxqgi3j82kns-gemfile-and-lockfile/Gemfile"
ENV["BUNDLE_PATH"] = "/nix/store/bg5mkks6rq2imp792yyz3zbkgiz6cwyf-mastodon-gems/lib/ruby/gems/2.6.0"
ENV['BUNDLE_FROZEN'] = '1'

Gem.use_paths("/nix/store/dpwih2mskka72049rypw539xg6ih2zqn-bundler-1.17.3/lib/ruby/gems/2.6.0", ENV["GEM_PATH"])

APP_PATH = File.expand_path('/nix/store/8x81l89rc9wz6i6qbj0pw4352safwwsn-mastodon-3.0.1/config/application', __dir__)
require_relative '/nix/store/8x81l89rc9wz6i6qbj0pw4352safwwsn-mastodon-3.0.1/config/boot'
require_relative '/nix/store/8x81l89rc9wz6i6qbj0pw4352safwwsn-mastodon-3.0.1/lib/cli'
Mastodon::CLI.start(ARGV)
```

petabyteboy replied:
I have to correct myself there: The wrappedRuby thing works, but in your code tootctl is not touched since it does not appear in mastodon-gems. I have committed a version that should fix all commands.

alyssais commented on pkgs/servers/mastodon/update.nix:
```diff
@@ -0,0 +1,39 @@
+    src = fetchFromGitHub {
+      owner = "petabyteboy";
+      repo = "bundix";
+      rev = "feature/submodules";
```
Should be pinned to a SHA (or, preferably, just patch upstream Bundix with `fetchpatch`). As it stands, this is not reproducible.

petabyteboy replied:
This will be gone very soon anyways

petabyteboy replied:
gone

alyssais commented on pkgs/servers/mastodon/update.nix:
```diff
@@ -0,0 +1,39 @@
+    };
+  });
+
+in writeScript "update-mastodon" ''
```
This is quite a long shell script to be embedded inside Nix. Consider an `update-mastodon.sh` and using `substituteAll` so it can be edited as a shell script more easily.

petabyteboy replied:
I changed it to a .sh using nix-shell as shebang, which should solve this issue.

petabyteboy commented on nixos/modules/services/web-apps/mastodon.nix:
```diff
@@ -0,0 +1,339 @@
+        '';
+        type = lib.types.str;
+      };
+      dbPassFile = lib.mkOption {
```
I will change the database options to match this proposal by @aanderse : https://github.com/NixOS/nixpkgs/issues/62073

happy-river commented on pkgs/servers/mastodon/default.nix:
```diff
@@ -0,0 +1,77 @@
+{ nodejs-slim, yarn2nix-moretea, fetchFromGitHub, bundlerEnv,
+  stdenv, yarn, lib, callPackage, ... }:
```
```suggestion
  stdenv, yarn, lib, callPackage, imagemagick, ffmpeg, file ... }:
```
These need to be added here since they are used below.

petabyteboy replied:
done

happy-river commented:
I've written [a second NixOS test](https://github.com/happy-river/nixpkgs/tree/mastodon-test) which runs Mastodon's Ruby and Javascript test suites.  Feel free to incorporate it into this pull request.

petabyteboy commented:
The tests seem to succeed :)

happy-river commented on pkgs/servers/mastodon/default.nix:
```diff
@@ -0,0 +1,84 @@
+    ln -s /var/log/mastodon log
+    ln -s /tmp tmp
+  '';
+  propagatedBuildInputs = [ imagemagick ffmpeg file ];
```
```suggestion
  propagatedBuildInputs = [ imagemagick ffmpeg file mastodon-gems.wrappedRuby ];
```
This makes the wrapped ruby executable available to `patchShebangs`.

petabyteboy commented on pkgs/servers/mastodon/default.nix:
```diff
@@ -0,0 +1,84 @@
+      if [ -f bin/$b ]; then
+        sed -i 's#/usr/bin/env ruby#${mastodon-gems.wrappedRuby}/bin/ruby#g' bin/$b
+      fi
+    done
```
Ah, that's a much nicer solution, thanks!

petabyteboy commented:
I started refactoring the database options and also adding createLocally options for smtp (postfix) and redis, with the overall goal of reducing the number of necessary options to set for the average setup, thus making it more user-friendly.

Unfortunately I found it very hard to find information on the mastodon configuration variables, and I couldn't figure out how to use a unix socket for postgresql yet.

happy-river commented:
> Unfortunately I found it very hard to find information on the mastodon configuration variables, and I couldn't figure out how to use a unix socket for postgresql yet.
> This prevents me from completing the database.createLocally implementation because we can not automatically create a user with a password.

`nixos/tests/mastodon/webserver.nix` uses postgresql with a socket.  I'm yet not convinced the configuration in that test is correct, but it gets the web server running.  I've resorted to searching the Mastodon source code to figure out what its configuration variables do.

petabyteboy commented:
> `nixos/tests/mastodon/webserver.nix` uses postgresql with a socket. I'm yet not convinced the configuration in that test is correct, but it gets the web server running. I've resorted to searching the Mastodon source code to figure out what its configuration variables do.

Actually the postgres socket thing works exactly like you did it in the test :sparkles:

aanderse commented on nixos/modules/services/databases/postgresql.nix:
```diff
@@ -22,6 +22,16 @@ let
       ${cfg.extraConfig}
     '';
 
+  ensureDatabasesOpts = {
```
@petabyteboy you might be interested in joining our [discussion](https://github.com/NixOS/nixpkgs/pull/68353#issuecomment-541994707) about whether extending the `mysql` and `postgresql` database provisioning options is appropriate for NixOS or not.

petabyteboy replied:
I see the point of that discussion, but to be honest I don't care. I don't need this functionality, I just added it because people asked for it and I will happily remove it when there's a consensus that we don't want this kind of functionality.

petabyteboy commented:
... that was me trying a button without a label and clicking it twice because the UI didn't respond.

So far I think I have addressed all comments other than comments that contradict eachother (discussions on other topics). From my side this is ready to go. Sure, I can probably work on this for another 5 months, but this is good enough to be included in nixpkgs in my view.

aanderse reviewed (commented):
I have lost track... what was the decision on `node` and `yarn`? Introduced back into nixpkgs?

Out of people who are following this thread and have commit access: are there (m)any remaining holdups on this?

I'll have to insist at this point the modifications are dropped on `postgresql` simply because there is a larger conversation happening that isn't resolved. Sorry to be a hassle.

Also... in the future keep in mind that you have been **exceptionally** accommodating to many people's requests in this PR. Someone stating a feature they think would be nice is not the same thing as a review that is critiquing code that has problems associated with it, or required missing features that you have overlooked. So don't feel bad to push back of you don't see value in a feature because after all it is you doing the work. :smile:

petabyteboy commented:
> I have lost track... what was the decision on `node` and `yarn`? Introduced back into nixpkgs?

yarn2nix was introduced back into nixpkgs as yarn2nix-moretea in https://github.com/NixOS/nixpkgs/pull/60429. node2nix was never really mentioned in this discussion, rather there were multiple competing projects with the name yarn2nix by Profpatsch and moretea. 
There were concerns about the maintainability of yarn2nix-moretea in nixpkgs, because in the past we had lost track of upstream. As I understand this is now solved by an update script.

> 
> Out of people who are following this thread and have commit access: are there (m)any remaining holdups on this?
> 
> I'll have to insist at this point the modifications are dropped on `postgresql` simply because there is a larger conversation happening that isn't resolved. Sorry to be a hassle.

Okay, I will do the following:
- I will create a branch on my fork to preserve the current state with postgresql changes.
- I will remove the postgresql changes and all functionality that depends on it on this branch, even though some people will be unhappy about using the migrations on an empty database

> 
> Also... in the future keep in mind that you have been **exceptionally** accommodating to many people's requests in this PR. Someone stating a feature they think would be nice is not the same thing as a review that is critiquing code that has problems associated with it, or required missing features that you have overlooked. So don't feel bad to push back of you don't see value in a feature because after all it is you doing the work.

Except some people do state their (hard to fulfill) wishes as if they were requirements, quoting a comment from this PR: "i would feel bad about merging this if there isn't first-class support for custom forks of mastodon"

petabyteboy commented on nixos/modules/services/web-apps/mastodon.nix:
```diff
@@ -0,0 +1,463 @@
+        '';
+      };
+
+      automaticMigrations = lib.mkOption {
```
It's not like it's complicated to add, so why not

petabyteboy commented on nixos/modules/services/web-apps/mastodon.nix:
```diff
@@ -0,0 +1,463 @@
+      };
+    };
+
+    services.postfix.enable = cfg.smtp.createLocally && cfg.smtp.host == "127.0.0.1";
```
Thanks!

petabyteboy commented on nixos/modules/services/web-apps/mastodon.nix:
```diff
@@ -0,0 +1,463 @@
+    };
+
+    services.postfix.enable = cfg.smtp.createLocally && cfg.smtp.host == "127.0.0.1";
+    services.redis.enable = cfg.redis.createLocally && cfg.redis.host == "127.0.0.1";
```
Thanks!

petabyteboy commented:
I have removed the postgresql changes and everything that depended on it. The version with those changes can be found in the branch feature/mastodon-postgres on my fork.

petabyteboy commented:
Note that the webserver test will not pass without the changes in the postgresql module, because manual intervention is required to initialize the database.

petabyteboy commented:
Actually it does work by using the migration chain, it's just quite slow.

ashkitten commented:
to clarify my wishes about custom forks -  i meant that more as a blanket statement about making it easy to update mastodon to any release, custom fork or not, because it seemed difficult to assemble all the required components and potentially easy to miss something. i was just advocating for an update script, really. sorry for confusion

happy-river commented:
> Actually it does work by using the migration chain, it's just quite slow.

It's too slow for the amount of patience I have, so I found a way to test whether a postgresql database has no tables yet and use `rake db:schema:load` instead of `rake db:migrate` in that case.  I've also ported the two mastodon tests to the new Python test driver.  Both commits can be found in my [mastodon-python-tests](https://github.com/happy-river/nixpkgs/tree/mastodon-python-tests) branch, which is based on this pull request rebased onto a recent master.

happy-river commented on nixos/modules/services/web-apps/mastodon.nix:
```diff
@@ -0,0 +1,467 @@
+          mkdir -p $(dirname ${cfg.vapidPrivateKeyFile}) $(dirname ${cfg.vapidPublicKeyFile})
+          keypair=$(bin/rake webpush:generate_keys)
+          echo $keypair | grep Private | sed 's/^Private -> //' > ${cfg.vapidPrivateKeyFile}
+          echo $keypair | grep Public | sed 's/^Public  -> //' > ${cfg.vapidPublicKeyFile}
```
This isn't working as expected.  With a configuration that doesn't specify the keys, it puts both keys on a single line into both `/var/lib/mastodon/secrets/vapid-private-key` and `vapid-public-key`:

```
Generated VAPID keypair: Public -> BC8xMKgr9Vtx9Sw2RHxnC__9d61l308waEeFS9XRyhtbbH8m5dsUyQ1S-2rxjEyvAfnOdf8AmzbDLHbAu-fCv4g= Private -> Noj-8wTsdT8zSY1ooBpbWGLJ3IPLj4WWkbSYUKikTr8=
```

petabyteboy replied:
Should be fixed, please confirm

happy-river replied:
Yes, it's fixed.

petabyteboy commented:
I have pushed your changes to my branch. Thanks!

happy-river commented on pkgs/top-level/all-packages.nix:
```diff
@@ -1886,6 +1886,8 @@ in
 
   massren = callPackage ../tools/misc/massren { };
 
+  mastodon = callPackage ../servers/mastodon { };
```
I recently realized that `all-packages.nix` is supposed to be sorted by directory.  Obviously lots of other people have missed that too. But to avoid adding to the disorder, Mastodon's proper place in the Servers section is down around line 15145, just before `mattermost`.

petabyteboy replied:
Thanks, I have moved it to the appropriate section

petabyteboy commented on pkgs/servers/mastodon/default.nix:
```diff
@@ -0,0 +1,79 @@
+  pname = "mastodon";
+  inherit src version;
+
+  passthru.updateScript = callPackage ./update.nix {};
```
Fixed.

happy-river commented on nixos/modules/services/web-apps/mastodon.nix:
```diff
@@ -0,0 +1,475 @@
+        VAPID_PRIVATE_KEY=$(cat ${cfg.vapidPrivateKeyFile})
+        VAPID_PUBLIC_KEY=$(cat ${cfg.vapidPublicKeyFile})
+        DB_PASS=$(cat ${cfg.database.passwordFile})
+        SMTP_PASSWORD=$(cat ${cfg.smtp.passwordFile})
```
To support passwords with spaces:
```
        DB_PASS="$(cat ${cfg.database.passwordFile})"
        SMTP_PASSWORD="$(cat ${cfg.smtp.passwordFile})"

```

happy-river commented on nixos/modules/services/web-apps/mastodon.nix:
```diff
@@ -0,0 +1,475 @@
+                join pg_namespace s on s.oid = c.relnamespace \
+                where s.nspname not in ('pg_catalog', 'pg_toast', 'information_schema') \
+                and s.nspname not like 'pg_temp%';" | sed -n 3p` -eq 0 ]; then
+          SAFETY_ASSURED=1 rake db:schema:load
```
```suggestion
          SAFETY_ASSURED=1 rake db:schema:load
          rake db:seed
```
Mastodon 3.0.0 [added an instance actor account](https://github.com/tootsuite/mastodon/pull/11321l) which must be seeded into a newly created production database.

happy-river commented on nixos/modules/services/web-apps/mastodon.nix:
```diff
@@ -0,0 +1,475 @@
+          Path to file containing the secret key base
+
+          Can be generated by running
+          cd $(nix-instantiate --eval '&lt;nixpkgs&gt;' -A mastodon.outPath); bin/rake secret
```
`nix-instantiate --eval '<nixpkgs>' -A mastodon.outPath` just tells you what the path of the Mastodon package is. It doesn't actually build the package, so if the package isn't already in the Nix store, as might be the case if you're just getting started setting up this module, this command won't work.

The best replacement I've come up with is: `nix build -f '&lt;nixpkgs&gt;' mastodon; cd result; bin/rake secret`

See my branch [60788-tootctl](https://github.com/happy-river/nixpkgs/tree/60788-tootctl) for a commit where I've made this change in the several places where it is needed, added mention that the keys will be generated automatically if they are not supplied, and also edited most of the other option descriptions for formatting and clarity.

happy-river commented:
I've now successfully created a Mastodon instance using this module and NixOps and a configuration that looks like this:

```
  tooter = { config, pkgs, ... }: {
    services.mastodon = {
      enable = true;
      configureNginx = true;
      localDomain = "example.com";
      smtp = {
        host = "smtp.example.org";
        port = 587;
        user = "user@mail.example.com";
        fromAddress = "Administrator <admin@mail.example.com>";
      };
    };
    networking.firewall.allowedTCPPorts = [ 80 443 ];
  };
```
I deployed it, copied my SMTP password over, restarted it, used its web interface to create an account, and then needed to use `tootctl` to promote the account to an administrator.  But `tootctl` needs Mastodon's environment variables to be set up correctly, or it won't work.  My solution to that problem can be found in my [60788-tootctl branch](https://github.com/happy-river/nixpkgs/tree/60788-tootctl). It creates a shell script `mastodon-env` which sets up the environment for another command.  With that done, making an administrator account became:

```
$ nixops ssh -d mastodon tooter
[root@tooter:~]# su - mastodon -s /bin/sh
[mastodon@tooter:~]$ mastodon-env tootctl accounts modify myaccount --role admin
```

This works, but I'm open to better solutions.  One feature that might be added to `mastodon-env` is to make it optionally `cd` to the Mastodon package directory, which would be helpful if you want to use it to run any `rake` tasks.

petabyteboy commented:
This project has caused me too much stress and I'm not really interested in improving it any further. Maybe @happy-river can open a PR and continue this.

Miaourt commented:
Don't worry, Take care of you @petabyteboy !

happy-river commented:
Thanks for everything you've done on this @petabyteboy ! I will continue to work on this and will create a new PR in a few days.

volth commented:
Any progress with this? It seems like a very good start.

bqv commented:
I actually got this working with a few monkeypatches. Been running it for almost a month. Seems like it shouldn't need much more work.

volth commented on pkgs/servers/mastodon/source-unpatched.nix:
```diff
@@ -0,0 +1,9 @@
+let
+  pkgs = import <nixpkgs> {};
```
I am not using `<nixpkgs>` variable at all (there is `<nixpkgs-booted>`, `<nixpkgs-current>` and `<nixpkgs-next>` instead), so this pattern does not work.
Probably better to change to
```
{ fetchFromGitHub }:
fetchFromGitHub {
 ...
}
```
and `callPackage ./source-unpatched.nix {};` instead of `import ./source-unpatched.nix;`;
It is even shorter.

happy-river replied:
I did pretty much the same thing in #78810 to remove the reference to `<nixpkgs>`.

happy-river commented:
I've just created #78810 which contains this pull request with improvements.
